### DISCUSSION TITLE
Error while running multiprocessing Queue in Gramine: No such file or directory.

### DISCUSSION THREAD

**Discussion by Smart-Yanmeng** [gramineproject/gramine#1363] [General]
I met an error while I running a program in Gramine, I hope someone can answer it for me.
This is my mainfest file:
```
loader.entrypoint = "file:{{ gramine.libos }}"
libos.entrypoint = "{{ entrypoint }}"

loader.log_level = "debug"

loader.env.LD_LIBRARY_PATH = "/lib:/usr/lib:{{ arch_libdir }}:/usr/{{ arch_libdir }}"

loader.env.OMP_NUM_THREADS = "8"

loader.insecure__use_cmdline_argv = true
loader.insecure__use_host_env = true

sgx.nonpie_binary = true

fs.mounts = [
  { uri = "file:{{ gramine.runtimedir() }}", path = "/lib" },
  { uri = "file:{{ arch_libdir }}", path = "{{ arch_libdir }}" },
  { uri = "file:/usr", path = "/usr" },
  { uri = "file:/etc", path = "/etc" },
  { uri = "file:{{ pillow_path }}", path = "{{ pillow_path }}" },

  { type = "tmpfs", path = "/tmp" },
]

sgx.enclave_size = "1G"
sgx.max_threads = 32

sgx.edmm_enable = {{ 'true' if env.get('EDMM', '0') == '1' else 'false' }}
sgx.trusted_files = [
  "file:{{ gramine.libos }}",
  "file:{{ entrypoint }}",
  "file:{{ gramine.runtimedir() }}/",
  "file:{{ arch_libdir }}/",
  "file:/usr/{{ arch_libdir }}/",
  "file:{{ python.stdlib }}/",
  "file:{{ python.distlib }}/",
  "file:{{ pillow_path }}",
  "file:{{ python.get_path('stdlib', vars={'installed_base': '/usr/local'}) }}/",

  "file:test.py",
]
```
This is my program file:
```
import multiprocessing

def worker(queue):
    item = queue.get()
    print("Worker received:", item)

if __name__ == "__main__":
    queue = multiprocessing.Queue()

    process = multiprocessing.Process(target=worker, args=(queue,))
    process.start()

    item = "Hello, multiprocessing!"
    queue.put(item)

    process.join()

    while not queue.empty():
        item = queue.get()
        print("Item in queue:", item)
```
And this is error:
```
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    queue = multiprocessing.Queue()
  File "/usr/lib/python3.8/multiprocessing/context.py", line 103, in Queue
    return Queue(maxsize, ctx=self.get_context())
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 42, in __init__
    self._rlock = ctx.Lock()
  File "/usr/lib/python3.8/multiprocessing/context.py", line 68, in Lock
    return Lock(ctx=self.get_context())
  File "/usr/lib/python3.8/multiprocessing/synchronize.py", line 162, in __init__
    SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
  File "/usr/lib/python3.8/multiprocessing/synchronize.py", line 57, in __init__
    sl = self._semlock = _multiprocessing.SemLock(
FileNotFoundError: [Errno 2] No such file or directory
```

**Comment by kailun-qin** [comment]
Hi @Smart-Yanmeng!

It looks like `_multiprocessing.SemLock` is implemented on top of the POSIX shared memory, see:
- https://github.com/python/cpython/blob/3.8/Modules/_multiprocessing/semaphore.c
- https://github.com/python/cpython/blob/3.8/Modules/_multiprocessing/posixshmem.c

This is currently unsupported (yet) in Gramine (see https://gramine.readthedocs.io/en/latest/devel/features.html#semaphores).

Pls kindly note that there is ongoing effort about adding this support: https://github.com/gramineproject/gramine/issues/757 and https://github.com/gramineproject/gramine/pull/827, which is currently planned for our next release (v1.5, probably in late May or early June). You can give it another try once we have it. Thanks!

**Reply by dimakuv** [reply]
> Pls kindly note that there is ongoing effort about adding this support: https://github.com/gramineproject/gramine/issues/757 and https://github.com/gramineproject/gramine/pull/827, which is currently planned for our next release (v1.5, probably in late May or early June). You can give it another try once we have it. Thanks!

I would like to point out that the "Shared Memory support" in Gramine is NOT secure, and thus should NOT be used for cases like this (enabling POSIX semaphores). This will lead to an insecure execution of your Python application, which is surely NOT what you want.

But as @kailun-qin points out, Python's `multiprocessing` module uses POSIX semaphores, and these are currently not supported in Gramine.

**Reply by Smart-Yanmeng** [reply]
Thanks for your reply! Maybe I should find another way to solve this problem. Thanks again!
